ci: Ensure SLSA Lvl 3 (#904)
kubewarden/kubewarden-controller: README.md

The [documentation](https://docs.kubewarden.io) provides more insights
about how the project works and how to use it.

# Software bill of materials & provenance

Kubewarden controller has its software bill of materials (SBOM) and build
[Provenance](https://slsa.dev/spec/v1.0/provenance) information published every
release. It follows the [SPDX](https://spdx.dev/) format and

> [SLSA](https://slsa.dev/provenance/v0.2#schema) provenance schema respectfully.

Reviewer: ```suggestion
[SLSA](https://slsa.dev/provenance/v0.2#schema) provenance schema.
```

Both of the files are generated by [Docker
buildx](https://docs.docker.com/build/metadata/attestations/) during the build
process and stored in the container registry together with the container image
as well as upload in the release page.

You can find them together with the signature and certificate used to sign it
in the [release
assets](https://github.com/kubewarden/kubewarden-controller/releases), and
attached to the image as JSON-encoded documents following the [in-toto SPDX
predicate](https://github.com/in-toto/attestation/blob/main/spec/predicates/spdx.md)
format. You can obtain them with
[`crane`](https://github.com/google/go-containerregistry/blob/main/cmd/crane/README.md)
or [`docker buildx imagetools
inspect`](https://docs.docker.com/reference/cli/docker/buildx/imagetools/inspect).

You can verify the container image with:

```shell
cosign verify-blob --certificate-oidc-issuer=https://token.actions.githubusercontent.com  \

>     --certificate-identity-regexp="https://github.com/kubewarden/kubewarden-controller*"  \

Reviewer 2: This example needs to be fully qualified, otherwise users may be checking against the wrong identity.

    --bundle kubewarden-controller-attestation-amd64-provenance-cosign.bundle \
    kubewarden-controller-attestation-amd64-provenance.json
```

> To verify verify the attestation manifest and its layers signatures:

Reviewer: Is that double plural accurate? Maybe 'layer signatures'?
```suggestion
To verify the attestation manifest and its layer signatures:
```

```shell
cosign verify --certificate-oidc-issuer=https://token.actions.githubusercontent.com  \

>     --certificate-identity-regexp="https://github.com/kubewarden/kubewarden-controller*" \

Reviewer 2: ditto

Author: Applied.

  "size": 1655,
  "annotations": {
    "vnd.docker.reference.digest": "sha256:611d499ec9a26034463f09fa4af4efe2856086252d233b38e3fc31b0b982d369",
    "vnd.docker.reference.type": "attestation-manifest"
  },
  "platform": {
    "architecture": "unknown",
    "os": "unknown"
  }
}
{
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "digest": "sha256:e0cd736c2241407114256e09a4cdeef55eb81dcd374c5785c4e5c9362a0088a2",
  "size": 1655,
  "annotations": {
    "vnd.docker.reference.digest": "sha256:03e5db83a25ea2ac498cf81226ab8db8eb53a74a2c9102e4a1da922d5f68b70f",
    "vnd.docker.reference.type": "attestation-manifest"
  },
  "platform": {
    "architecture": "unknown",
    "os": "unknown"
  }
}
```

Then you can use the `digest` field to verify the attestation manifest and its
layers signatures.

```shell
cosign verify --certificate-oidc-issuer=https://token.actions.githubusercontent.com  \
    --certificate-identity-regexp="https://github.com/kubewarden/kubewarden-controller*" \
    ghcr.io/kubewarden/kubewarden-controller@sha256:fc01fa6c82cffeffd23b737c7e6b153357d1e499295818dad0c7d207f64e6ee8

> crane manifest  ghcr.io/kubewarden/kubewarden-controller@sha256:fc01fa6c82cffeffd23b737c7e6b153357d1e499295818dad0c7d207f64e6ee8                                                                                                  ✔ │ 14:28:50

Reviewer 3: Ops, let's remove the time from my zsh theme. xD

Author: Woops, done! :sweat_smile:

    },
    {
      "mediaType": "application/vnd.in-toto+json",
      "digest": "sha256:7ce0572628290373e17ba0bbb44a9ec3c94ba36034124931d322ca3fbfb768d9",
      "size": 7363045,
      "annotations": {
        "in-toto.io/predicate-type": "https://spdx.dev/Document"
      }
    },
    {
      "mediaType": "application/vnd.in-toto+json",
      "digest": "sha256:dacf511c5ec7fd87e8692bd08c3ced2c46f4da72e7271b82f1b3720d5b0a8877",
      "size": 71331,
      "annotations": {
        "in-toto.io/predicate-type": "https://spdx.dev/Document"
      }
    },
    {
      "mediaType": "application/vnd.in-toto+json",
      "digest": "sha256:594da3e8bd8c6ee2682b0db35857933f9558fd98ec092344a6c1e31398082f4d",
      "size": 980,
      "annotations": {
        "in-toto.io/predicate-type": "https://spdx.dev/Document"
      }
    },
    {
      "mediaType": "application/vnd.in-toto+json",
      "digest": "sha256:7738d8d506c6482aaaef1d22ed920468ffaf4975afd28f49bb50dba2c20bf2ca",
      "size": 13838,
      "annotations": {
        "in-toto.io/predicate-type": "https://slsa.dev/provenance/v0.2"
      }
    }
  ]
}

cosign verify --certificate-oidc-issuer=https://token.actions.githubusercontent.com  \

>     --certificate-identity-regexp="https://github.com/kubewarden/kubewarden-controller*" \

Reviewer 2: Ditto

Author: Applied.

    ghcr.io/kubewarden/kubewarden-controller@sha256:594da3e8bd8c6ee2682b0db35857933f9558fd98ec092344a6c1e31398082f4d
```

Note that each attestation manifest (for each architecture) has its own layers.

> Each layers is a different SBOM SPDX or provenance files generated by Docker

Reviewer: Possible unnecessary pluralization?
```suggestion
Each layer is a different SBOM SPDX or provenance file generated by Docker
```
